Detectors/EMCAL/reconstruction/src/CaloRawFitterStandard.cxx

```diff
@@ -0,0 +1,143 @@
+{
+  float time = 0;
+  float amp = 0;
+  float chi2 = 0;
+  int ndf = 0;
+  bool fitDone = kFALSE;
+
+  auto [nsamples, bunchIndex, ampEstimate,
+        maxADC, timeEstimate, pedEstimate, first, last] = preFitEvaluateSamples(bunchlist, altrocfg1, altrocfg2, mAmpCut);
+
+  if (ampEstimate >= mAmpCut) {
+    time = timeEstimate;
+    int timebinOffset = bunchlist.at(bunchIndex).getStartTime() - (bunchlist.at(bunchIndex).getBunchLength() - 1);
+    amp = ampEstimate;
+
+    if (nsamples > 1 && maxADC < constants::OVERFLOWCUT) {
+      auto [amp, time, chi2, fitDone] = fitRaw(first, last);
+      time += timebinOffset;
+      timeEstimate += timebinOffset;
+      ndf = nsamples - 2;
+    }
+  }
+  if (fitDone) {
+    float ampAsymm = (amp - ampEstimate) / (amp + ampEstimate);
+    float timeDiff = time - timeEstimate;
+
+    if ((TMath::Abs(ampAsymm) > 0.1) || (TMath::Abs(timeDiff) > 2)) {
+      amp = ampEstimate;
+      time = timeEstimate;
+      fitDone = kFALSE;
+    }
+  }
+  if (amp >= mAmpCut) {
+    if (!fitDone) {
+      std::default_random_engine generator;
+      std::uniform_real_distribution<float> distribution(0.0, 1.0);
+      amp += (0.5 - distribution(generator));
+    }
+    time = time * constants::EMCAL_TIMESAMPLE;
+    time -= mL1Phase;
+
+    return CaloFitResults(-99, -99, mAlgo, amp, time,
+                          (int)time, chi2, ndf);
+  }
+  return CaloFitResults(-1, -1);
+}
+
+std::tuple<float, float, float, bool> CaloRawFitterStandard::fitRaw(int firstTimeBin, int lastTimeBin) const
+{
+
+  float amp(0), time(0), chi2(0);
+  bool fitDone(false);
+
+  int nsamples = lastTimeBin - firstTimeBin + 1;
+  fitDone = kFALSE;
+  if (nsamples < 3)
+    return std::make_tuple(amp, time, chi2, fitDone);
+
+  TGraph gSig(nsamples);
+
+  for (int i = 0; i < nsamples; i++) {
+    int timebin = firstTimeBin + i;
+    gSig.SetPoint(i, timebin, getReversed(timebin));
+  }
+
+  TF1 signalF("signal", CaloRawFitterStandard::rawResponseFunction, 0, constants::EMCAL_MAXTIMEBINS, 5);
+
+  signalF.SetParameters(10., 5., constants::TAU, constants::ORDER, 0.); //set all defaults once, just to be safe
+  signalF.SetParNames("amp", "t0", "tau", "N", "ped");
+  signalF.FixParameter(2, constants::TAU);
+  signalF.FixParameter(3, constants::ORDER);
+  signalF.FixParameter(4, 0);
+  signalF.SetParameter(1, time);
+  signalF.SetParameter(0, amp);
+  signalF.SetParLimits(0, 0.5 * amp, 2 * amp);
+  signalF.SetParLimits(1, time - 4, time + 4);
+
+  try {
+    gSig.Fit(&signalF, "QROW"); // Note option 'W': equal errors on all points
+    amp = signalF.GetParameter(0);
+    time = signalF.GetParameter(1);
+    chi2 = signalF.GetChisquare();
+    fitDone = kTRUE;
+  } catch (const std::exception& e) {
+    LOG(ERROR) << "CaloRawFitterStandard Fit exception " << e.what();
+    // stay with default amp and time in case of exception, i.e. no special action required
+    fitDone = kFALSE;
+  }
```

Comment:
Not sure this try-catch block is necessary and useful, to my knowledge ROOT's fitting methods don't throw exceptions. Failed fits must be handled differently.